File: backend/interfaces/api/v1/voice.py
from typing import Any

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from application.voice.voice_service import VoiceService
from interfaces.dependencies import get_voice_service

router = APIRouter(prefix="/api/v1", tags=["voice"])


def success_response(data):
    return {"code": 0, "message": "success", "data": data}


class ExtractRequest(BaseModel):
    model_config = {"extra": "forbid"}

    texts: list[str]
    name: str = "default"
    novel_id: str = ""


class DriftRequest(BaseModel):
    model_config = {"extra": "forbid"}

    baseline_id: str
    sample_text: str


class RewriteRequest(BaseModel):
    model_config = {"extra": "forbid"}

    baseline_id: str
    target_text: str
    drift_dimensions: list[str] = []


@router.post("/voice/extract")
def extract_fingerprint(
    request: ExtractRequest,
    service: VoiceService = Depends(get_voice_service),
):
    fp = service.extract_fingerprint(
        texts=request.texts,
        name=request.name,
        novel_id=request.novel_id,
    )
    return success_response(fp.to_dict())


@router.get("/voice/fingerprints")
def list_fingerprints(
    service: VoiceService = Depends(get_voice_service),
):
    fps = service.list_fingerprints()
    return success_response(fps)


@router.get("/voice/fingerprints/{fp_id}")
def get_fingerprint(
    fp_id: str,
    service: VoiceService = Depends(get_voice_service),
):
    fp = service.get_fingerprint(fp_id)
    if not fp:
        raise HTTPException(status_code=404, detail="Fingerprint not found")
    return success_response(fp.to_dict())


@router.post("/voice/detect-drift")
async def detect_drift(
    request: DriftRequest,
    service: VoiceService = Depends(get_voice_service),
):
    result = service.detect_drift(
        baseline_id=request.baseline_id,
        sample_text=request.sample_text,
    )
    if not result:
        raise HTTPException(status_code=404, detail="Baseline fingerprint not found")
    return success_response(result.to_dict())


@router.post("/voice/rewrite-prompt")
def generate_rewrite_prompt(
    request: RewriteRequest,
    service: VoiceService = Depends(get_voice_service),
):
    prompt = service.generate_rewrite_prompt(
        baseline_id=request.baseline_id,
        target_text=request.target_text,
        drift_dimensions=request.drift_dimensions,
    )
    if not prompt:
        raise HTTPException(status_code=404, detail="Baseline fingerprint not found")
    return success_response({"prompt": prompt})


@router.get("/voice/fingerprints/{fp_id}/style-guide")
def get_style_guide(
    fp_id: str,
    service: VoiceService = Depends(get_voice_service),
):
    guide = service.generate_style_guide(fp_id)
    if not guide:
        raise HTTPException(status_code=404, detail="Fingerprint not found")
    return success_response({"style_guide": guide})


@router.post("/voice/fingerprints/merge")
def merge_fingerprints(
    request: ExtractRequest,
    service: VoiceService = Depends(get_voice_service),
):
    fps: list[Any] = []
    for text in request.texts:
        fp = service.extract_fingerprint(
            texts=[text],
            name=f"sample_{len(fps)}",
            novel_id=request.novel_id,
        )
        fps.append(fp)

    merged = service.merge_fingerprints(fps, name=request.name)
    return success_response(merged.to_dict())

File: frontend.stellar.bak/src/api/voice.ts
import request from './http'

export interface VoiceFingerprint {
  fingerprint_id: string
  novel_id: string
  name: string
  lexical_richness: number
  sentence_length_mean: number
  sentence_length_std: number
  paragraph_length_mean: number
  paragraph_length_std: number
  unique_char_ratio: number
  function_word_ratio: number
  content_word_ratio: number
  dialogue_ratio: number
  narration_ratio: number
  description_ratio: number
  punctuation_density: Record<string, number>
  common_ngrams: string[]
  signature_phrases: string[]
  sentence_structure_diversity: number
  paragraph_starts: string[]
  avg_word_per_sentence: number
  clause_density: number
  source_sample_count: number
  source_char_count: number
  version: number
  created_at: string
  updated_at: string
}

export interface VoiceDriftResult {
  drifted: boolean
  overall_similarity: number
  dimension_scores: Record<string, number>
  drift_dimensions: string[]
  details: Record<string, unknown>
}

export interface VoiceFingerprintInfo {
  id: string
  name: string
  novel_id: string
  sample_count: number
  char_count: number
}

export function extractFingerprint(
  texts: string[],
  name: string = 'default',
  novelId: string = ''
) {
  return request.post<VoiceFingerprint>('/voice/extract', { texts, name, novel_id: novelId })
}

export function listFingerprints() {
  return request.get<VoiceFingerprintInfo[]>('/voice/fingerprints')
}

export function getFingerprint(fpId: string) {
  return request.get<VoiceFingerprint>(`/voice/fingerprints/${fpId}`)
}

export function detectDrift(baselineId: string, sampleText: string) {
  return request.post<VoiceDriftResult>('/voice/detect-drift', {
    baseline_id: baselineId,
    sample_text: sampleText,
  })
}

export function generateRewritePrompt(
  baselineId: string,
  targetText: string,
  driftDimensions: string[] = []
) {
  return request.post<{ prompt: string }>('/voice/rewrite-prompt', {
    baseline_id: baselineId,
    target_text: targetText,
    drift_dimensions: driftDimensions,
  })
}

export function getStyleGuide(fpId: string) {
  return request.get<{ style_guide: string }>(`/voice/fingerprints/${fpId}/style-guide`)
}

export function mergeFingerprints(texts: string[], name: string, novelId: string = '') {
  return request.post<VoiceFingerprint>('/voice/fingerprints/merge', {
    texts,
    name,
    novel_id: novelId,
  })
}
